source/common/http/alternate_protocols_cache_impl.h

```diff
@@ -0,0 +1,43 @@
+#pragma once
+
+#include "envoy/common/backoff_strategy.h"
+#include "envoy/thread_local/thread_local.h"
+
+#include "common/http/alternate_protocols_cache.h"
+
+#include "absl/container/flat_hash_map.h"
+
+namespace Envoy {
+namespace Http {
+
+class AlternateProtocolsCacheImpl : public AlternateProtocolsCache {
+public:
+  AlternateProtocolsCacheImpl(ThreadLocal::Instance& tls, TimeSource& time_source);
+  ~AlternateProtocolsCacheImpl() override;
+
+  // AlternateProtocols
+  void setAlternatives(const Origin& origin, const std::vector<AlternateProtocol>& protocols,
+                       const MonotonicTime& expiration) override;
+  OptRef<const std::vector<AlternateProtocol>> findAlternatives(const Origin& origin) override;
+  size_t size() const override;
+
+private:
+  struct Entry {
+    std::vector<AlternateProtocol> protocols_;
+    MonotonicTime expiration_;
+  };
+
+  struct State : ThreadLocal::ThreadLocalObject {
+    // Map from hostname to list of alternate protocols.
+    // TODO(RyanTheOptimist): Add a limit to the size of this map and evict based on usage.
+    std::map<Origin, Entry> protocols_;
+  };
+
+  // Time source used to check expiration of entries.
+  TimeSource& time_source_;
+  // Thread local state for the cache
+  ThreadLocal::TypedSlot<State> slot_;
```

Comment:
What's the tradeoff between using thread-local storage vs a lock here? My intuition would be that a lock-based approach would increase QUIC usage and require less memory, but that could be my client-side habits.

Reply:
This is the approach taken by the DNS cache which alyssa suggested we use for alt-svc as well. My experience has been to avoid locks basically at all costs. But I'm happy to do something different if that makes more sense. For the client-side use case, I'd be somewhat surprised if we ran with multiple networking threads so that might avoid the additional memory, but I haven't thought about this deeply.